internal/common/eventstream/eventbatcher.go

```diff
@@ -0,0 +1,121 @@
+package eventstream
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/G-Research/armada/pkg/api"
+)
+
+type eventBatchCallback func(events []*api.EventMessage) error
+
+type EventBatcher interface {
+	Register(callback eventBatchCallback)
+	Report(event *api.EventMessage) error
+	Flush() error
+	Stop()
+}
+
+type TimedEventBatcher struct {
+	batchSize int
+	timeout   time.Duration
+
+	batch    []*api.EventMessage
+	callback eventBatchCallback
+
+	mutex sync.Mutex
+	t     *time.Timer
+}
+
+func NewTimedEventBatcher(batchSize int, timeout time.Duration) *TimedEventBatcher {
+	b := &TimedEventBatcher{
+		batchSize: batchSize,
+		timeout:   timeout,
+		batch:     []*api.EventMessage{},
+	}
+	return b
+}
+
+func (b *TimedEventBatcher) Register(callback eventBatchCallback) {
+	b.callback = callback
+}
+
+func (b *TimedEventBatcher) Report(event *api.EventMessage) error {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	b.batch = append(b.batch, event)
+	if len(b.batch) >= b.batchSize {
+		err := b.flushUnlocked()
+		if err != nil {
+			return err
+		}
+	} else {
+		if b.t != nil {
+			b.t.Stop()
+		}
+		b.t = time.AfterFunc(1*time.Second, func() {
```

Comment:
My inexperience in Go is probably showing here but does this mean that every event will trigger a Flush, albeit one that happens a second after the event is reported and that may just try and flush am empty buffer?

If this is the case it sounds a bit hard to reason about and potentially a bit inefficient.  Do you think it would be possible to flush when the buffer is full or a second after the last message